PowerEditor/src/WinControls/ProjectPanel/DirectoryWatcher.cpp

```diff
@@ -0,0 +1,389 @@
+	{
+		_forcedUpdateToAdd.insert(hItem);
+		update();
+	}
+}
+
+void DirectoryWatcher::forceUpdateAll()
+{
+	Yuni::MutexLocker lock(_lock);
+	for (auto it=_dirItems.begin(); it !=_dirItems.end(); ++it)
+	{
+		_forcedUpdateToAdd.insert(it->first);
+	}
+	update();
+}
+
+void DirectoryWatcher::update()
+{
+	::SetEvent(_hUpdateEvent);
+}
+
+void DirectoryWatcher::startThread()
+{
+
+	try {
+		Yuni::MutexLocker lock(_lock);
+		if (_running)
+			return;
+
+		_hRunningEvent = ::CreateEvent(NULL, TRUE, FALSE, NULL);
+		_hStopEvent = ::CreateEvent(NULL, TRUE, FALSE, NULL);
+		_hUpdateEvent = ::CreateEvent(NULL, TRUE, FALSE, NULL);
+		if (!_hStopEvent || !_hRunningEvent || !_hUpdateEvent)
+			throw std::runtime_error("Could not create DirectoryWatcher events");
+
+		_hThread = ::CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)threadFunc, (LPVOID)this, 0, NULL);
+		if (!_hThread)
+			throw std::runtime_error("Could not create DirectoryWatcher thread");
+
+		::WaitForSingleObject(_hRunningEvent, INFINITE);
+	}
+	catch (...)
+	{
+		if (_hRunningEvent)
+			::CloseHandle(_hRunningEvent);
+		if (_hStopEvent)
+			::CloseHandle(_hStopEvent);
+		if (_hUpdateEvent)
+			::CloseHandle(_hUpdateEvent);
+		throw;
+	}
+
+
+
+}
+
+void DirectoryWatcher::stopThread()
+{
+	Yuni::MutexLocker lock(_lock);
+
+	if (_running)
+	{
+		::SetEvent(_hStopEvent);
+		::WaitForSingleObject(_hThread, INFINITE);
+		::CloseHandle(_hThread);
+		::CloseHandle(_hRunningEvent);
+		::CloseHandle(_hStopEvent);
+		_running = false;
+		_hThread = NULL;
+		_hRunningEvent = NULL;
+		_hStopEvent = NULL;
+	}
+
+}
+
+int DirectoryWatcher::thread()
+{
+
+	_running = true;
+	::SetEvent(_hRunningEvent);
+
+	HANDLE events[2];
+	events[0] = _hStopEvent;
+	events[1] = _hUpdateEvent;
+
+	::SetThreadPriority( ::GetCurrentThread(), THREAD_PRIORITY_LOWEST );
+
+	for(;;)
+	{
+		updateDirs();
+		iterateDirs();
+		DWORD waitresult = WaitForMultipleObjects(2, events, FALSE, _manualMode ? INFINITE : _updateFrequencyMs);
+		switch (waitresult)
+		{
+			case WAIT_FAILED:			// error
+				return -1;
+			case WAIT_OBJECT_0:			// stop event
+				ResetEvent(_hStopEvent);
+				return 0;
+			case WAIT_OBJECT_0+1:		// update instantly
+				ResetEvent(_hUpdateEvent);
+				break;
+			case WAIT_TIMEOUT:			// timeout - normal polling frequency
+				break;
+			default:					// should not happen
+				assert(0);
+				break;
+		}
+
+	}
+
+}
+
+DWORD DirectoryWatcher::threadFunc(LPVOID data)
+{
+	DirectoryWatcher* pw = static_cast<DirectoryWatcher*>(data);
+	return (DWORD)pw->thread();
+
+}
+
+bool DirectoryWatcher::post(HTREEITEM item, UINT message)
+{
+	if (!_hWnd)
+		return false;
+
+	if (message == DIRECTORYWATCHER_UPDATE)
+		_changeOccurred = true;
+
+	LRESULT smResult = SendMessageTimeout(_hWnd, message, 0, (LPARAM)item, SMTO_ABORTIFHUNG, 10000, NULL);
+	return smResult != 0;
+}
+
+// main function to iterate the directories.
+void DirectoryWatcher::iterateDirs()
+{
+
+	std::map<Directory*,bool> changedMap;
```

Comment:
Does this map need to be ordered? `std::unordered_map` can be *a __lot faster__*.